demos/guide/quic-server-non-block.c

```diff
@@ -0,0 +1,479 @@
+     * whether or how the client authenticated the server, so the term "mutual"
+     * is potentially misleading).
+     *
+     * Since we're not soliciting or processing client certificates, we don't
+     * need to configure a trusted-certificate store, so no call to
+     * SSL_CTX_set_default_verify_paths() is needed.  The server's own
+     * certificate chain is assumed valid.
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_NONE, NULL);
+
+    /* Setup ALPN negotiation callback to decide which ALPN is accepted. */
+    SSL_CTX_set_alpn_select_cb(ctx, select_alpn, NULL);
+
+    return ctx;
+
+err:
+    SSL_CTX_free(ctx);
+    return NULL;
+}
+
+/* Create UDP socket on the given port. */
+static int create_socket(uint16_t port)
+{
+    int fd;
+    struct sockaddr_in sa = {0};
+
+    /* Retrieve the file descriptor for a new UDP socket */
+    if ((fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
+        fprintf(stderr, "cannot create socket");
+        return -1;
+    }
+
+    sa.sin_family = AF_INET;
+    sa.sin_port = htons(port);
+
+    /* Bind to the new UDP socket on localhost */
+    if (bind(fd, (const struct sockaddr *)&sa, sizeof(sa)) < 0) {
+        fprintf(stderr, "cannot bind to %u\n", port);
+        BIO_closesocket(fd);
+        return -1;
+    }
+
+    /* Set port to nonblocking mode */
+    if (BIO_socket_nbio(fd, 1) <= 0) {
+        fprintf(stderr, "Unable to set port to nonblocking mode");
+        BIO_closesocket(fd);
+        return -1;
+    }
+
+    return fd;
+}
+
+/**
+ * @brief Waits for activity on the SSL socket, either for reading or writing.
+ *
+ * This function monitors the underlying file descriptor of the given SSL
+ * connection to determine when it is ready for reading or writing, or both.
+ * It uses the select function to wait until the socket is either readable
+ * or writable, depending on what the SSL connection requires.
+ *
+ * @param ssl A pointer to the SSL object representing the connection.
+ *
+ * @note This function blocks until there is activity on the socket. In a real
+ * application, you might want to perform other tasks while waiting, such as
+ * updating a GUI or handling other connections.
+ *
+ * @note This function uses select for simplicity and portability. Depending
+ * on your application's requirements, you might consider using other
+ * mechanisms like poll or epoll for handling multiple file descriptors.
+ */
+static void wait_for_activity(SSL *ssl)
+{
+    int sock, isinfinite;
+    fd_set read_fd;
+    struct timeval tv;
+    struct timeval *tvp = NULL;
+
+    /* Get hold of the underlying file descriptor for the socket */
+    if ((sock = SSL_get_fd(ssl)) == -1) {
+        fprintf(stderr, "Unable to get file descriptor");
+        return;
+    }
+
+    /* Initialize the fd_set structure */
+    FD_ZERO(&read_fd);
+
+    /* Add the socket file descriptor to the fd_set */
+    FD_SET(sock, &read_fd);
+
```

Comment:
This still doesn't look quite right to me. I'd really expect this function to be identical to the quic client non-block version. In particular the calls to `SSL_net_read_desired` and `SSL_net_write_desired` are missing here.

Reply:
When I add those calls, the program crashes. This may be a bug with our QUIC implementation

```diff
diff --git a/demos/guide/quic-server-non-block.c b/demos/guide/quic-server-non-block.c
index 4ad4e7b361..e222e5ed9a 100644
--- a/demos/guide/quic-server-non-block.c
+++ b/demos/guide/quic-server-non-block.c
@@ -206,7 +206,7 @@ static int create_socket(uint16_t port)
 static void wait_for_activity(SSL *ssl)
 {
     int sock, isinfinite;
-    fd_set read_fd;
+    fd_set read_fd, write_fd;
     struct timeval tv;
     struct timeval *tvp = NULL;
 
@@ -218,6 +218,16 @@ static void wait_for_activity(SSL *ssl)
 
     /* Initialize the fd_set structure */
     FD_ZERO(&read_fd);
+    FD_ZERO(&write_fd);
+
+    /*
+     * Find out if we would like to write to the socket, or read from it (or
+     * both)
+     */
+    if (SSL_net_write_desired(ssl))
+        FD_SET(sock, &write_fd);
+    if (SSL_net_read_desired(ssl))
+        FD_SET(sock, &read_fd);
 
     /* Add the socket file descriptor to the fd_set */
     FD_SET(sock, &read_fd);
```

Running with LLDB:

```console
$ lldb ./quic-non-blocking-server 4443 chain.pem pkey.pem
(lldb) target create "./quic-server-non-block"
Current executable set to 'openssl/demos/guide/quic-server-non-block' (arm64).
(lldb) settings set -- target.run-args  "4443" "chain.pem" "pkey.pem"
(lldb) r
Process 35602 launched: 'openssl/demos/guide/quic-server-non-block' (arm64)
Waiting for connection
Process 35602 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xa0)
    frame #0: 0x000000010023630c libssl.3.dylib`ossl_quic_get_net_write_desired + 68
libssl.3.dylib`ossl_quic_get_net_write_desired:
->  0x10023630c <+68>: ldr    x0, [x8, #0xa0]
    0x100236310 <+72>: bl     0x100227cb4    ; ossl_quic_channel_get_reactor
    0x100236314 <+76>: bl     0x100245848    ; ossl_quic_reactor_net_write_desired
    0x100236318 <+80>: mov    x8, x0
Target 0: (quic-server-non-block) stopped.
```

Reply:
Ah ok, so this is because we are the QUIC channel in `ssl` is not yet initiated yet (since this is called before there is a connection). So maybe we can call these functions only if a connection has already been established

Reply:
This definitely looks like a bug in our QUIC implementation to me. We should not crash in this circumstance (I would expect `SSL_net_write_desired()` or `SSL_net_read_desired()` should return 0 if there is no connection yet.

Reply:
Oh....wait is this because we are calling it on the listener?? This should also work (probably `SSL_net_read_desired()` should always return 1 in this case??)

Reply:
Fix for this crash in #25642. With that fix applied this no longer crashes - so we should remove the code you added here to check the read state

Reply:
Ok thanks, I've removed the code to check the read state: [`b49df8c` (#25431)](https://github.com/openssl/openssl/pull/25431/commits/b49df8c984a4ab9cf9acdcff4da11b146c19fe11)